### DISCUSSION TITLE
Description list with multiple terms per definition and multiple definitions per term

### DISCUSSION THREAD

**Discussion by dimztimz** [jgm/djot#309] [Ideas]
The most common case for description list is to have list of pairs of [one term and one description/definition](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl#single_term_and_description). But HTML supports other scenarios, namely:

- [Multiple terms, single description](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl#multiple_terms_single_description);
- [Single term, multiple descriptions](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl#single_term_multiple_descriptions).

Seems to me to support such scenario, a change in the syntax is needed. I just noticed this discussion #193 for another syntax.

**Comment by dereuromark** [comment]
I looked into that

### Multiple terms → single description is more commonly needed in practice.

  1. Technical documentation: Abbreviations + full names
  : API
  : Application Programming Interface

    A set of protocols for building software...
  2. Glossaries: Synonyms/aliases
  : directory
  : folder

    A container for files in a filesystem.
  3. Multilingual docs: Same concept, different terms
  : color
  : colour

    The visual property of objects...

### The other pattern (one term → multiple descriptions) is more niche:

  - Mainly useful for dictionaries where words have distinct meanings
  - Most technical docs don't need this - they just use separate paragraphs within one <dd>

### POC  
I created a demo:
https://github.com/php-collective/djot-php/pull/11

The syntax is intuitive (consecutive : term lines) and covers the 80% use case. The implementation change is also minimal - just collect terms before hitting an indented block.

The other direction is more complicated and as such I didnt look further into that 20% use case.

**Reply by dereuromark** [reply]
See https://github.com/jgm/djot/issues/128 discussion for follow up on multiple dd.

```
: Djot
 (/dʒɑt/)
  A lightweight markup language designed by John MacFarlane,
  creator of Pandoc and CommonMark.
: +
  Enhanced by Mark Scherer and Djot PHP implementation to support more awesomeness.

: Markdown
: CommonMark
: GFM

  Earlier markup languages that inspired Djot's design,
  but with less consistent parsing rules.

  They can be up to two times slower.
```

Live demo of it: https://sandbox.dereuromark.de/sandbox/djot/complex-examples

**Reply by dereuromark** [reply]
Note: In [Carve](https://markup-carve.github.io/carve/) I took a clearer path:

```
:: Carve
:  A lightweight markup language with clean syntax.
:  Secure by design

:: Markdown
:: Djot
:  The predecessors that inspired Carve.
```

(The blank line between groups is purely visual - remove it and the output is identical.)

This renders to:

```html
<dl>
  <dt>Carve</dt>
  <dd>A lightweight markup language with clean syntax.</dd>
  <dd>Secure by design</dd>
  <dt>Markdown</dt>
  <dt>Djot</dt>
  <dd>The predecessors that inspired Carve.</dd>
</dl>
```

So one group can carry multiple terms (`<dt>`) and multiple descriptions (`<dd>`) - the full shape HTML's `<dl>` actually allows. For targets that permit only a single term per group, the terms can be merged into one `<dt>` as multiple lines/paragraphs - a rendering choice, in my view.

Maybe something similar could be adopted in Djot.

**Comment by jgm** [comment]
Personally I don't think multiple terms or multiple definitions are needed in light markup.  For multiple terms, you can always use two terms separated by a hard line break, comma, or whatever. For multiple definitions, you can use e.g. an ordered list for the different definitions.

```
: apple

  1. red fruit
  2. nice computer
```

**Reply by bpj** [reply]
My take is that they may be nice to have, and if I were to design a full LML without needing to be backward compatible with anything else I would borrow the MediaWiki markers: `;` for term and `:` for definitions, but if possible allowing both single line

````code
;term1: def1
;term2: def2
````

and multi-line syntax

````code
;   Term1\: line 1.
    Line 2 of term 1.
:   Def 1 line1
    def 1 line 2.

;   Term2a
;    Term2b
:    Def 2
````

Not that I have thought through all details. I might support both fenced groups and inline groups similar to Lua long strings to "keep things together" without corresponding to any outline markup to work around single/multi-line issues in my LML!
